### PR TITLE
[CL20] Fix data race in Device execution test suite

### PR DESCRIPTION
**Pull request by alexbatashev**
The following kernels ran into data races:
- enqueue_flags_wait_work_group_simple
- enqueue_flags_wait_work_group_event
- enqueue_flags_wait_work_group_local
- enqueue_flags_wait_work_group_event_local
Fix KhronosGroup#40

### CODE REVIEW

File: test_conformance/device_execution/enqueue_flags.cpp [hunk 1]
```diff
@@ -39,6 +39,8 @@ static const char* enqueue_flags_wait_kernel_simple[] =
     NL, "  size_t lid = get_local_id(0);"
     NL, "  size_t tid = get_global_id(0);"
     NL, ""
+    NL, "  array[index * gs + tid] = array[(index - 1) * gs + tid] + 1;"
```
Comment (b-sumner) [inline test_conformance/device_execution/enqueue_flags.cpp]: Why was this moved?  The whole point of this test is to ensure the child waits for the parent.

File: test_conformance/device_execution/enqueue_flags.cpp [hunk 2]
```diff
@@ -50,11 +52,9 @@ static const char* enqueue_flags_wait_kernel_simple[] =
     NL, "    }"
     NL, "  }"
     NL, ""
-    NL, "  array[index * gs + tid] = array[(index - 1) * gs + tid] + 1;"
```
Comment (b-sumner) [inline test_conformance/device_execution/enqueue_flags.cpp]: This location is correct.

File: test_conformance/device_execution/enqueue_flags.cpp [hunk 3]
```diff
@@ -82,7 +82,6 @@ static const char* enqueue_flags_wait_kernel_simple[] =
     NL, "  size_t gs  = get_global_size(0);"
     NL, "  size_t tid  = get_global_id(0);"
     NL, ""
-    NL, "  res[tid] = 0;"
     NL, "  array[tid] = tid;"
```
Comment (b-sumner) [inline test_conformance/device_execution/enqueue_flags.cpp]: It would actually be better to move this after the enqueue.

Comment (alexbatashev) [reply test_conformance/device_execution/enqueue_flags.cpp]: This will not resolve the data race condition.

File: test_conformance/device_execution/enqueue_flags.cpp (hunk 1 above)
Comment (alexbatashev) [reply test_conformance/device_execution/enqueue_flags.cpp]: done

File: test_conformance/device_execution/enqueue_flags.cpp (hunk 2 above)
Comment (alexbatashev) [reply test_conformance/device_execution/enqueue_flags.cpp]: done

File: test_conformance/device_execution/enqueue_flags.cpp (hunk 3 above)
Comment (b-sumner) [reply test_conformance/device_execution/enqueue_flags.cpp]: Agreed, but it would improve the test.

Comment (alexbatashev) [reply test_conformance/device_execution/enqueue_flags.cpp]: Ok, done

File: test_conformance/device_execution/enqueue_flags.cpp [hunk 4]
```diff
@@ -185,7 +183,7 @@ static const char* enqueue_flags_wait_kernel_local[] =
     NL, "  size_t tid = get_global_id(0);"
     NL, ""
     NL, "  sub_array[lid] = array[(index - 1) * gs + tid];"
-    NL, "  barrier(CLK_LOCAL_MEM_FENCE);"
+    NL, "  barrier(CLK_GLOBAL_MEM_FENCE);"
```
Comment (b-sumner) [inline test_conformance/device_execution/enqueue_flags.cpp]: The local fence is needed here so all threads in the loop below see all the updates

File: test_conformance/device_execution/enqueue_flags.cpp [hunk 5]
```diff
@@ -236,7 +234,6 @@ static const char* enqueue_flags_wait_kernel_local[] =
     NL, "  size_t gs  = get_global_size(0);"
     NL, "  size_t tid  = get_global_id(0);"
     NL, ""
-    NL, "  res[tid] = 0;"
     NL, "  array[tid] = tid;"
```
Comment (b-sumner) [inline test_conformance/device_execution/enqueue_flags.cpp]: It would improve the test to move this after the enqueue

File: test_conformance/device_execution/enqueue_flags.cpp [hunk 6]
```diff
@@ -262,7 +259,7 @@ static const char* enqueue_flags_wait_kernel_event_local[] =
     NL, "  size_t tid = get_global_id(0);"
     NL, ""
     NL, "  sub_array[lid] = array[(index - 1) * gs + tid];"
-    NL, "  barrier(CLK_LOCAL_MEM_FENCE);"
+    NL, "  barrier(CLK_GLOBAL_MEM_FENCE);"
```
Comment (b-sumner) [inline test_conformance/device_execution/enqueue_flags.cpp]: The local fence is required here.

File: test_conformance/device_execution/enqueue_flags.cpp [hunk 7]
```diff
@@ -499,7 +495,7 @@ static const char* enqueue_flags_wait_work_group_local[] =
     NL, "  size_t gs = get_global_size(0);"
     NL, ""
     NL, "  sub_array[lid] = array[(index - 1) * gs + tid];"
-    NL, "  barrier(CLK_LOCAL_MEM_FENCE);"
+    NL, "  barrier(CLK_GLOBAL_MEM_FENCE);"
```
Comment (b-sumner) [inline test_conformance/device_execution/enqueue_flags.cpp]: The local fence is required here.

File: test_conformance/device_execution/enqueue_flags.cpp [hunk 8]
```diff
@@ -580,7 +575,7 @@ static const char* enqueue_flags_wait_work_group_event_local[] =
...
```
Comment (b-sumner) [inline test_conformance/device_execution/enqueue_flags.cpp]: The local fence is required here

File: test_conformance/device_execution/enqueue_flags.cpp (hunk 4 above)
Comment (alexbatashev) [reply test_conformance/device_execution/enqueue_flags.cpp]: Ah, sorry. It's been a while, I completely forgot there were working barriers

**Comment by alexbatashev** [comment]
@b-sumner do you have any complaints about latest revision?

**Comment by alexbatashev** [comment]
@kpet do these changes make sense to you?

File: test_conformance/device_execution/enqueue_flags.cpp [hunk 9]
```diff
@@ -676,7 +672,7 @@ int test_enqueue_flags(cl_device_id device, cl_context context, cl_command_queue
     cl_uint i;
     cl_int err_ret, res = 0;
     clCommandQueueWrapper dev_queue;
-    cl_int kernel_results[MAX_GWS] = { -1 };
+    cl_int kernel_results[MAX_GWS] = { 0 };
```
Comment (kpet) [inline test_conformance/device_execution/enqueue_flags.cpp]: Why did you change that line? My understanding of the validation logic below is that both 0 and 1 are expected values in the output. Isn't it better to initialise the results array with a value that can never be valid? I'm afraid this might mask failures.

Comment (alexbatashev) [reply test_conformance/device_execution/enqueue_flags.cpp]: The value is being overwritten in kernel anyway. But the way some kernels re-initialize res array causes data races. So, after some discussions here and there I decided to change this value to 0. This should not seriously affect the test. But definitely fixes data races.

Comment (kpet) [reply test_conformance/device_execution/enqueue_flags.cpp]: Not sure I'm following. Don't we want to guarantee that the entirety of the array is filled by the kernels? It seems to me that we may be hiding issues in the kernels with this change.

>This should not seriously affect the test. But definitely fixes data races.

I can't reconcile these two statements :).

Comment (alexbatashev) [reply test_conformance/device_execution/enqueue_flags.cpp]: The original issue #40 contains detailed explanations with pictures https://github.com/KhronosGroup/OpenCL-CTS/issues/40#issuecomment-456696470 

This is the approach we agreed on with @b-sumner offline. I'll add you to the mail thread.
